src/context.rs

```diff
@@ -0,0 +1,230 @@
+use futures::channel::oneshot;
+use futures::future::{pending, select, BoxFuture, FutureExt, Shared};
+use tokio::sync::mpsc;
+use tokio::time::{delay_for, Duration};
+
+pub struct CancelHandle {
+    cancel_fn: Box<dyn FnOnce()>,
+}
+
+impl CancelHandle {
+    pub fn cancel(self) {
+        let _ = (self.cancel_fn)();
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Context {
+    pub done: Shared<BoxFuture<'static, ()>>,
+}
+
+impl Context {
+    pub fn new() -> Self {
+        Self {
+            done: pending().boxed().shared(),
+        }
+    }
+
+    pub fn with_cancel(&self) -> (Self, CancelHandle) {
+        let (cancel_tx, cancel_rx) = oneshot::channel::<()>();
+        let parent_done = self.done.clone();
+        let done = select(parent_done, cancel_rx).map(|_| ()).boxed().shared();
+        let cancel_fn = Box::new(move || {
+            let _ = cancel_tx.send(());
+        });
+        (Self { done }, CancelHandle { cancel_fn })
+    }
+
+    pub fn with_timeout(&self, timeout: Duration) -> (Self, CancelHandle) {
+        let (mut cancel_tx, mut cancel_rx) = mpsc::channel::<()>(2);
+        let cancel_fn = Box::new(move || {
+            let _ = cancel_tx.send(());
+        });
+
+        let parent_done = self.done.clone();
+        let done = async move {
+            tokio::select! {
+                _ = parent_done => (),
+                _ = delay_for(timeout) => (),
+                _ = cancel_rx.recv() => (),
+            }
```

Comment:
Tokio has a [`timeout`](https://docs.rs/tokio/0.2.22/tokio/time/fn.timeout.html) future wrapper as well.

Reply:
Oh that's nice, yeah, I think we can just wrap this API